File: python/Q207_course_schedule.py
#
# @lc app=leetcode id=207 lang=python3
#
# [207] Course Schedule
#

# @lc code=start
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegrees: list[int] = [0] * numCourses
        adjecent_list: dict[int, list[int]] = {i: [] for i in range(numCourses)}

        for i, j in prerequisites:
            adjecent_list[j].append(i)
            indegrees[i] += 1

        leaves: list[int] = [i for i, v in enumerate(indegrees) if v == 0]

        finished: int = 0
        while leaves:
            finished += 1
            node: int = leaves.pop()

            for n in adjecent_list[node]:
                indegrees[n] -= 1
                if indegrees[n] == 0:
                    leaves.append(n)

        return finished == numCourses


# @lc code=end

File: rust/src/leetcode/Q207_course_schedule.rs
/*
 * @lc app=leetcode id=207 lang=rust
 *
 * [207] Course Schedule
 */

// @lc code=start
impl Solution {
    pub fn can_finish(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> bool {
        let mut indegrees: Vec<usize> = vec![0; num_courses as usize];
        let mut prerequisites_map: Vec<Vec<usize>> = vec![vec![]; num_courses as usize];

        for prerequisite in prerequisites {
            indegrees[prerequisite[0] as usize] += 1;
            prerequisites_map[prerequisite[1] as usize].push(prerequisite[0] as usize);
        }

        let mut current: Vec<usize> = indegrees
            .iter()
            .enumerate()
            .filter_map(|(course, &indegree)| (indegree == 0).then_some(course))
            .collect();

        let mut finished: i32 = 0;
        while !current.is_empty() {
            finished += 1;
            let course: usize = current.pop().unwrap();

            for &advanced_course in &prerequisites_map[course] {
                indegrees[advanced_course] -= 1;
                if indegrees[advanced_course] == 0 {
                    current.push(advanced_course);
                }
            }
        }

        finished == num_courses
    }
}
// @lc code=end
